client/src/main/java/de/hellfish/ngrok/client/NgrokClient.java

```diff
@@ -60,27 +84,66 @@ public void run() {
         }
         running = initConnectionWithServer();
         while (running) {
-                try {
-                    HttpRequest userRequest = HttpRequestSerializer.readFromInputStream(serverSocket.getInputStream());
-                    log.info("Client received new user request: {}", userRequest.toString());
-                } catch (IOException e) {
-                    log.error("Server is down");
-                    running = false;
-                }
+            processRequest();
         }
         stop();
     }
 
+    private void processRequest() {
+        try {
+            userRequest = HttpRequestSerializer.readFromInputStream(serverSocket.getInputStream());
+            log.info("Client received new user request: {}", userRequest.toString());
+        } catch (IOException e) {
+            log.error("Error reading user request from server", e);
+            running = false;
+        }
+        if (userRequest == null) {
+            return;
+        }
+
+        HttpResponse<String> serviceResponse = null;
+        try {
+            serviceSocket = new Socket("localhost", servicePort);
+            HttpClient httpClient = HttpClient.newHttpClient();
+            serviceResponse = httpClient.send(convertHttpRequest(userRequest), HttpResponse.BodyHandlers.ofString());
+            log.info("Client got response from service: {} with body {}", serviceResponse, serviceResponse.body());
+        } catch (IOException | InterruptedException e) {
+            log.error("Error sending http-request to service", e);
+            sendResponse503ToServer();
+        }
```

Comment:
I suggest handling the IOException and InterruptedException exceptions in separate catch blocks. For InterruptedException, I propose using the Thread.currentThread().interrupt() method to restore the thread interruption flag. This will allow subsequent checks or tests to determine whether the thread was interrupted. If, of course, you will even need such a check in your future code